dali/kernels/any.h

```diff
@@ -0,0 +1,394 @@
+// Copyright (c) 2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_ANY_H_
+#define DALI_KERNELS_ANY_H_
+
+#include <exception>
+#include <utility>
+#include <type_traits>
+
+namespace dali {
+
+namespace detail {
+
+
+struct alignas(8)
+any_placeholder {
+  constexpr any_placeholder() = default;
+  char data[8] = {};
+};
+
+using any_ = void(any_placeholder *);
+
+struct any_helper_base {
+  virtual void destroy(any_placeholder *placeholder) const = 0;
+  virtual void free(any_placeholder *placeholder) const = 0;
+  virtual void *get_void(any_placeholder *placeholder) const noexcept = 0;
+  virtual const void *get_void(const any_placeholder *placeholder) const noexcept = 0;
+  virtual void clone(any_placeholder *dest, const any_placeholder *src) const = 0;
+  virtual void placement_clone(any_placeholder *dest, const any_placeholder *src) const = 0;
+  virtual bool is_same_ex(const any_helper_base *other) const = 0;
+
+  inline bool is_same(const any_helper_base *h) const {
+    return h && (this == h || is_same_ex(h));
+  }
+
+  template <typename T>
+  T *get(any_placeholder *p) const noexcept {
+    return reinterpret_cast<T*>(get_void(p));
+  }
+  template <typename T>
+  const T *get(const any_placeholder *p) const noexcept {
+    return reinterpret_cast<const T*>(get_void(p));
+  }
+};
+
+template <typename T, bool dynamic =
```

Comment:
Also, where are you using the second template arg?

Reply:
Only for partial specialization. I gave it a name to convey some meaning.